src/Metadata/PropertyMetadata.php

```diff
@@ -113,6 +113,11 @@ class PropertyMetadata extends BasePropertyMetadata
      */
     public $readOnly = false;
 
+    /**
+     * @var string
+     */
+    public $readOnlyIf = null;
```

Comment:
instead of adding one new property, we can re-use the previous one, that can be `bool|Expression`. In this way the exposed api will be simpler.

For annotations will be as you suggested `@ReadOnly(if="something")`, but for yaml and XML, we can have:

XML example:

- `read-only="true"`  as it was before
- `read-only="false"`  as it was before
- `read-only="some expression"` works as before but will be evaluated as expression.

In this way most of the works is done by the metadata mappers

Reply:
That was actually my first approach, but then I saw how it was done on exclude-if so I thought I should go for the same schema.
I will change that.
Using the same schema on exclude-if would cause a BC break though.